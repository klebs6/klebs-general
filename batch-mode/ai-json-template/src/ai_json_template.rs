// ---------------- [ File: ai-json-template/src/ai_json_template.rs ]
crate::ix!();

/// The derived code implements `AiJsonTemplate` for each struct, letting you
/// call `MyStruct::to_template()` to get a JSON “schema” describing how the
/// AI should produce data that matches this layout.
///
pub trait AiJsonTemplate
: Clone 
+ Debug 
+ LoadFromFile 
+ SaveToFile<Error=SaveLoadError> 
+ Serialize 
+ for<'a> Deserialize<'a> 
{
    /// Return a JSON template describing how the AI’s output should be structured.
    /// This might include doc comments or other instructions for each field.
    fn to_template() -> serde_json::Value;
}

/// Extended trait that demands both a normal `AiJsonTemplate` and
/// a `...Justification` structure for *nested* usage.
pub trait AiJsonTemplateWithJustification: AiJsonTemplate {
    fn to_template_with_justification() -> serde_json::Value;
}
