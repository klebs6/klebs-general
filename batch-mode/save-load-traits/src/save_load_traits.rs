// ---------------- [ File: src/save_load_traits.rs ]
crate::ix!();

#[async_trait]
pub trait SaveToFile {

    type Error;

    async fn save_to_file(
        &self,
        filename: impl AsRef<Path> + Send,

    ) -> Result<(), Self::Error>;
}

#[async_trait]
pub trait LoadFromFile: Sized {

    type Error;

    async fn load_from_file(filename: impl AsRef<Path> + Send) 
        -> Result<Self, Self::Error>;

}
