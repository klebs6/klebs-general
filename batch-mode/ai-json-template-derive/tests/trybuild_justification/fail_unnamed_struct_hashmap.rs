// ---------------- [ File: ai-json-template-derive/tests/trybuild_justification/fail_unnamed_struct_hashmap.rs ]
#![allow(dead_code)]
#![allow(unused_imports)]

use ai_json_template::*;
use ai_json_template_derive::*;
use std::collections::HashMap;
use serde::{Serialize, Deserialize};
use getset::{Getters, Setters};
use derive_builder::Builder;
use save_load_derive::*;
use save_load_traits::*;

/// A tuple struct => should fail
#[derive(
    SaveLoad,
    Debug,
    Clone,
    Serialize,
    Deserialize,
    AiJsonTemplate,
    AiJsonTemplateWithJustification,
)]
struct TupleStructMap(HashMap<String, bool>);

fn main() {}
