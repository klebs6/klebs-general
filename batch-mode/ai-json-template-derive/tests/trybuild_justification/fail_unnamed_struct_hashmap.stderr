error: AiJsonTemplate derive only supports named fields for structs (or handle your tuple struct in the enum logic).
  --> tests/trybuild_justification/fail_unnamed_struct_hashmap.rs:14:1
   |
14 | / /// A tuple struct => should fail
15 | | #[derive(
16 | |     SaveLoad,
17 | |     Debug,
...  |
23 | | )]
24 | | struct TupleStructMap(HashMap<String, bool>);
   | |_____________________________________________^

error: AiJsonTemplateWithJustification only supports named fields for structs.
  --> tests/trybuild_justification/fail_unnamed_struct_hashmap.rs:14:1
   |
14 | / /// A tuple struct => should fail
15 | | #[derive(
16 | |     SaveLoad,
17 | |     Debug,
...  |
23 | | )]
24 | | struct TupleStructMap(HashMap<String, bool>);
   | |_____________________________________________^
